client/state/reducer-registry/index.js

```diff
@@ -0,0 +1,152 @@
+			return reducer( initialState, { type: DESERIALIZE } );
+		},
+	} );
+}
+
+/**
+ * Add reducers into a combined reducer returned by combineReducersAndAddLater.
+ * @param {object} reducersToAdd - object containing the additional reducers to merge
+ * @param {string} name - Unique name to identify the reducer collection
+ * @param {object} registry - Registry to use, default to module variable
+ * @param {object} dispatch - Redux store.dispatch() method to call, default to module variable
+ */
+export const addReducers = function( reducersToAdd, name = 'root', registry = registeredReducers, dispatch = defaultStoreDispatch ) {
+	const reducers = registry[ name ];
+	if ( ! reducers ) {
+		throw new Error( `Reducer registry name ${ name } does not exist.` );
+	}
+	const keys = Object.keys( reducersToAdd ).filter( key => ! reducers[ key ] );
+	if ( ! keys.length ) {
+		return;
+	}
+
+	const reducersToSet = keys.map( key => {
+		const reducer = reducersToAdd[ key ];
+		const { schema, hasCustomPersistence } = reducer;
+		return {
+			key,
+			reducer: hasCustomPersistence ? reducer : withSchemaValidation( schema, reducer ),
+		};
+	} );
+
+	if ( dispatch ) {
+		// Store has initialized, we will need to deserialize the part.
+		reducersToSet.forEach( ( { key, reducer } ) => {
+			reducers[ key ] = getDeserializeReducer( reducer );
+		} );
+
+		dispatch( { type: DESERIALIZE_PART } );
+	}
+
+	reducersToSet.forEach( ( { key, reducer } ) => {
+		reducers[ key ] = reducer;
+	} );
+};
+
+/**
+ * combineReducersAndAddLater returns a reducer much like the one returned by
+ * combineReducers, with the exception that you may add into the reducers collection
+ * later. It also forgives undefined keys by calling placeholderReducer to handle
+ * these keys.
+ *
+ * This make the reducer function non-pure but it allow us to keep the unloaded
+ * states persistent, and swap the reducer when it loads.
+ *
+ * @param {object} initialReducers - object containing the reducers to merge
+ * @param {string} name - Unique name to identify this reducer collection
+ * @param {object} registry - Registry to use, default to module variable
+ * @returns {function} - Returns the combined reducer function
+ */
+export const combineReducersAndAddLater = function( initialReducers, name = 'root', registry = registeredReducers ) {
+	if ( registry[ name ] ) {
+		throw new Error( `Duplicate reducer registry name ${ name }.` );
+	}
+
+	const reducers = ( registry[ name ] = reduce(
+		initialReducers,
+		( validated, next, key ) => {
+			const { schema, hasCustomPersistence } = next;
+			return {
+				...validated,
+				[ key ]: hasCustomPersistence ? next : withSchemaValidation( schema, next ),
+			};
+		},
+		{}
+	) );
+
+	const reducer = function( state = {}, action ) {
+		const keys = uniq( Object.keys( reducers ).concat( Object.keys( state ) ) );
+		let hasChanged = false;
+		const nextState = {};
+		for ( let i = 0; i < keys.length; i++ ) {
```

Comment:
General note for readability that we prefer map/filter/forEach over for loops. We also try and avoid `let` if possible so there's less context needed to know what variable has what value.

Reply:
Ah some copy pasta from the original `combineReducers`? https://github.com/reactjs/redux/blob/master/src/combineReducers.js#L167

Reply:
Yes, this mimics the original `combineReducers` impl in Redux.

I assume keeping the impl simular helps the readability in some way? Also, `forEach` can be a little bit slower — I am not sure if we feel this place is critical enough to forgo some readability.

Let me know and I will update it.

Reply:
We can keep the faster implementation, but it's definitely worthy of a comment since it doesn't match our code styles.